### PR TITLE
St 48 dept back end

### PR DESCRIPTION
**Pull request by Garrett-Preston**
There is a problem with mocking the del() function of knex. Will discuss with Caleb tomorrow or tuesday. In the meantime, here is the code for the department backend.

### CODE REVIEW

File: tests/server_tests/department.test.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  select: jest.fn().mockReturnThis(),
+  where: jest.fn().mockReturnThis(),
+  from: jest.fn().mockReturnThis(),
+  andWhere: jest.fn().mockReturnThis(),
+  then: jest.fn().mockReturnValue([{ userId: 3, userName: 'test' }]),
+  join: jest.fn().mockReturnThis(),
+  insert: jest.fn().mockReturnThis(),
+  into: jest.fn().mockReturnThis(),
+  onConflict: jest.fn(async (callback) => callback()),
+  delete: jest.fn().mockReturnThis()
```
[the remark this answers is not in the document]
Comment (Blackkirby72) [reply tests/server_tests/department.test.js:30]: at the beginning of the test(s) for the delete route, just add something like
```javascript
jest.spyOn(dbClient, 'del').mockReturnValue(0)
```
make sure the return value is useful for what you test (I think the query normally returns the number of rows deleted?)

**Review by Blackkirby72 — COMMENTED** [review]
Ideally, we'd be passing and setting the authorization header around which has the current user's ID stored within (could also add their deptIDs in the future), so we wouldn't have to pass a *logged in user's ID* so long as we verify the auth token. That might have to come next sprint though

File: tests/server_tests/department.test.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  where: jest.fn().mockReturnThis(),
+  insert: jest.fn().mockReturnThis(),
+  select: jest.fn().mockReturnThis(),
+  from: jest.fn().mockReturnThis(),
+  into: jest.fn().mockReturnThis(),
+  andWhere: jest.fn().mockReturnThis(),
+  then: jest.fn().mockReturnThis(),
+  transacting: jest.fn().mockReturnThis()
+}
+
```
Comment (Blackkirby72) [inline tests/server_tests/department.test.js:18]: You don't need this if you aren't using transactions in the file you're testing

File: tests/server_tests/department.test.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+jest.mock('../../server/api/dbClient', () => ({
+  transaction: jest.fn(async (callback) => { await callback(mockTrx) }),
```
Comment (Blackkirby72) [inline tests/server_tests/department.test.js:20]: can also remove this line

**Comment by Garrett-Preston** [comment]
This needs some polish on top and I need to discuss the authorization token with Caleb - but for the purposes of demo and sprint review I'm submitting this PR

**Pull request closed without merging**